resolves #3 - Added auto skip for shorts with low likes

adsuth opened the pull request:
Added an additional section to the popup for "extra options".
I figured that this option would be too uncommon for it to be in the regular controls.

![image](https://github.com/ynshung/better-yt-shorts/assets/74009107/8ae1b081-4b7c-43f1-b401-e0bccce4c6f4)

There is a checkbox that enables this feature
Theres an input through which the user can set the threshold for the number of likes a post must have in order to be skipped.

Finally, theres a save button. This *could* just be done via events, but I figured thatd be far more costly - especially with the input.

potential issues:
- I havent tested on firefox (this is my first time editing extensions and i couldnt figure out how to open this in it)
- I wrote a method to convert formatted numbers like 1K to 1000. Not sure if other languages use other letters, if so, that'll need changed. (There must be a builtin for that im missing)

swolewizard commented:
Thank you for taking the time to code this unique feature. I've thoroughly tested it for an hour or two and encountered a few issues that I believe need attention.

**Bugs:**

1. The feature doesn't skip shorts with 0 likes.
2. It also fails to skip shorts that have been skipped previously in the same session.

**Requests:**

1. I would like to propose a new toggle option - 'Auto Don't recommend this channel if under X likes.' This addition aims to prevent situations where the same short is played twice, resulting in it not being skipped the second time.

Other than these mentioned issues, the feature works as intended, automatically skipping any video that has more than 1 like and is under the inputted value.

Your time and effort are greatly appreciated, and I'm looking forward to seeing these issues resolved and this feature merged. Please let me know if you need any further clarification or assistance in addressing these matters. Thank you!

adsuth commented:
Hi,

Thank you for your input and help with testing.

Regarding 0 likes, this slipped past but I know how to go about fixing it. 
Just need to check if the like count is "Like". If so, convert that to 0 (probably at this point in the code)
<img width="377" alt="image" src="https://github.com/ynshung/better-yt-shorts/assets/74009107/fc02b05e-bf62-45fd-85ba-8174c93401f0">

I'll get this sorted if and when I have time.

---
Regarding your request, I'm not sure of the feasibility. I don't know if active user actions can be automated in this way, though it'd certainly be worth exploring. The way this plugins

I suppose there may be a way to emulate this through saving a blacklist to local storage of any channel name that had a skipped short, but it'd have to be investigated.

The issue with this would be that, say a channel you like recently uploaded a short and it hasn't taken off yet. They would automatically get blocked from your feed which is definitely not desirable.

swolewizard commented:
> The issue with this would be that, say a channel you like recently uploaded a short and it hasn't taken off yet. They would automatically get blocked from your feed which is definitely not desirable.

Yeah I just realized this so I removed the request, you're right.

swolewizard commented:
I've tested the code by returning 0 if LikeCount is "Like"

```
const getLikeCount = (id) => {
  // there may be a more direct way of getting the actual value, but I spent 5 hours and this is the best I could find lmao
  const likesElement = document.querySelector(
    `[id='${id}']  > div.overlay.style-scope.ytd-reel-video-renderer > ytd-reel-player-overlay-renderer #like-button`
  );
  const numberOfLikes = likesElement.firstElementChild.innerText.split(/\r?\n/)[0];

  // Check if the like count is "Like" and convert it to 0
  if (numberOfLikes.toLowerCase() === "like") {
    return 0;
  }

  return convertLocaleNumber(numberOfLikes);
};
```

However, the problem arises when the short is initially loaded, and its LikeCount is always "Like," which updates to the proper LikeCount milliseconds later. As a result, this change skips the majority of videos. Additionally, there is sometimes a bug where videos with 1,600+ comments will have their LikeCount displayed as "Like," even though they clearly have hundreds of thousands of likes.

So, I'm not too sure about the approach I should take here; I'm stumped. I'll leave it to you for when you have time.

adsuth commented:
> However, the problem arises when the short is initially loaded, and its LikeCount is always "Like," which updates to the proper LikeCount milliseconds later. 

Added a fix, and also addresses this. It now shouldnt skip videos that havent started yet

---
Also incorporated a few other languages - other langs format numbers differently. eg "15K" in spanish is "15 K", and languages like german dont even use qualifiers like "K", and instead format the number like "15.000" (and yeah, some use commas, others dots.)

**Havent added them all, but the structure is there to support them.**
<img width="400" alt="image" src="https://github.com/ynshung/better-yt-shorts/assets/74009107/d3735500-f66a-45fd-b2e2-14cea4229cec">


---
Also, incorporated a way to accomodate other languages for the 0 likes thing. Obviously, not all languages say "Like", french for example uses "j'aime". 
**Again, only did a few, but its easy enough to add more.**
<img width="300" alt="image" src="https://github.com/ynshung/better-yt-shorts/assets/74009107/1b5806ab-99de-4a69-91f6-fbcb54972ded">

swolewizard commented:
Thanks for the quick update. I've been using this new version for about 30 minutes, and I encountered a few problems:

1. It skips every other random short regardless of the LikeCount, similar to my previous solution, where I just returned 0 if LikeCount was "Like". This issue occurs so frequently that I reverted back to the previous version. 

However, upon using the previous version, I encountered a new problem. 

2. Every time it skipped a video, I would scroll back up to verify if it skipped correctly based on the inputted minimum value. Approximately one-third of the time, it would skip a video above the minimum with 60k, 1.8m, 700k, etc. likes.

It seems there might be something wrong with the logic of the code or the way YouTube loads their shorts. Please look into this when you have time. I really appreciate the effort you've put into addressing these issues. Let me know if you need any further information. Thank you!

swolewizard commented:
@adsuth, I've submitted a pull request to your fork, and I believe I've resolved all the issues I've been experiencing. I'll be conducting thorough testing now, and in case I encounter any bugs, I'll let you know. Alternatively, if needed, I'll submit another pull request to address them myself.

ynshung commented on content-script.js:
```diff
@@ -433,6 +546,112 @@ const timer = setInterval(() => {
+  if ( hasMultiplier )
+    return +string.slice( 0, end ).replace( /,|\./g, "" ) * multipliers[ multiplier ]
+
+  return +string.slice( 0, end + 1 ).replace( /,|\./g, "" ) 
```
Wouldn't this cause inaccuracies in the final value? Namely if the like count is `5.6M` and the if the dot is removed it will be like multiplied by ten here.

swolewizard replied:
It would, replacing 
`string.slice(0, end) `

with 

`string.slice(0, end + 1)`

should fix it.

swolewizard replied:
nevermind that won't fix it, it'd actually return 0

swolewizard replied:
Fixed this issue here
https://github.com/adsuth/better-yt-shorts/pull/2#issue-1816827392

ynshung commented:
I've removed the comment check since there is a possibility that some user is skipping larger number of likes on the shorts video so comment checking wouldn't work there. Maybe can be another separate criteria but I believe checking likes would be enough for now.

ynshung commented:
Also, seems to be working fine in Firefox but the extra options might need to be worked on:

<img src="https://github.com/ynshung/better-yt-shorts/assets/61302840/83a41298-3d56-457e-a037-9f3772de264a" width="256"/>

adsuth commented:
issue with the change @swolewizard 
langs that dont use multipliers + use dots instead of commas as separators

eg, deutsch. This is treated as a double
<img width="78" alt="image" src="https://github.com/adsuth/better-yt-shorts/assets/74009107/c484caa5-bcae-4280-95c8-a1e56aa9704a">

i think an easy fix would be to just remove the dot if theres no multiplier, im not available to do it atm

swolewizard commented:
I managed to fix this issue, but during the process, I noticed that the German millions use a comma instead of a decimal point. It took me 1-2 hours to figure that out.

https://github.com/adsuth/better-yt-shorts/pull/2#issuecomment-1646686442

ynshung commented:
LGTM, I'm going ahead to merge and publish this but feel free to addon anything later, thanks @adsuth and @swolewizard for the help! 🎉
